llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,391 @@
+                                     "shouldn't contain empty metadata nodes");
+
+    MDString *TypeName = cast<MDString>(N.getOperand(0));
+    Type *T = StructType::getTypeByName(C, TypeName->getString());
+    assert(T &&
+           "invalid type referenced by intel_types_that_use_aspect metadata");
+
+    AspectsSetTy &Aspects = Result[T];
+    for (size_t I = 1; I != N.getNumOperands(); ++I) {
+      ConstantAsMetadata *CAM = dyn_cast<ConstantAsMetadata>(N.getOperand(I));
+      assert(CAM && "constant metadata is expected in "
+                    "intel_types_that_use_aspect list's entry");
+      Constant *C = CAM->getValue();
+      Aspects.insert(cast<ConstantInt>(C)->getSExtValue());
+    }
+  }
+
+  return Result;
+}
+
+using TypesEdgesTy =
+    std::unordered_map<const Type *, std::vector<const Type *>>;
+
+/// Propagates aspects from type @Start to all types which
+/// are reachable by edges @Edges by BFS algorithm.
+/// Result is recorded in @ResultAspects.
+void PropagateAspectsThroughTypes(const TypesEdgesTy &Edges, const Type *Start,
+                                  const AspectsSetTy &AspectsToPropagate,
+                                  TypeToAspectsMapTy &ResultAspects) {
+  SmallPtrSet<const Type *, 16> Visited;
+  std::queue<const Type *> queue;
+  queue.push(Start);
+  while (!queue.empty()) {
+    const Type *T = queue.front();
+    queue.pop();
+    if (Visited.count(T))
+      continue;
+
+    Visited.insert(T);
+    ResultAspects[T].insert(AspectsToPropagate.begin(),
+                            AspectsToPropagate.end());
+    if (!Edges.count(T)) {
+      continue;
+    }
+
+    for (const Type *TT : Edges.at(T)) {
+      if (Visited.count(TT))
+        continue;
+
+      queue.push(TT);
+    }
+  }
+}
+
+/// Returns all types with corresponding aspects from module @M. Type T in the
+/// result contains an aspect A if there is a way with instance of T to access
+/// by internal fields or pointers another type TT which uses corresponding
+/// aspect A.
+/// @TypesWithAspects argument consist of known types with aspects
+/// from metadata information.
+///
+/// The algorithm is the following:
+/// 1) Make a list of all structure types from module @M. The list also
+///    contains DoubleTy since it is optional as well.
+/// 2) Make from list a type graph which consist of nodes corresponding to types
+///    and directed edges between nodes. An edge from type A to type B
+///    corresponds to the fact that A is accessible from an instance of type B
+///    (by direct field or pointer in B).
+/// 3) For every known type with aspects propagate it's aspects over graph.
+///    Every propagation is a separate run of BFS algorithm.
+///
+/// Time complexity: O((V + E) * T) where T is the number of input types
+/// containing aspects.
+TypeToAspectsMapTy
+GetTypesWithAspectsFromModule(const Module &M,
+                              TypeToAspectsMapTy &TypesWithAspects) {
```

Comment:
It still looks weird to me that in order to get types with aspects from a module, we need to pass types with aspects.
What about renaming it into something like `PropagateAspectsToOtherTypesInModule` ?

Ideally, we shouldn't be even returning any values, because we can directly modify `TypesWithAspects` map which is passed to us, but that will require preserving the map keys into a temporary store before iterating over map (to avoid possible issues with inserting during iterating)

Reply:
Changed it.